### PR TITLE
Implement BO Ausschreibung

### LINKED ISSUE #192: Implement BO Ausschreibung
https://www.bo4e.de/dokumentation/geschaeftsobjekte/bo-ausschreibung
requires
- #156 
- #129 
- #94 
- #108 
-

### PR DESCRIPTION
**Pull request by hf-kklein**
Fixes #192

### CODE REVIEW

**Review by hf-fvesely — APPROVED** [review]

File: src/bo4e/bo/ausschreibung.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    # required attributes
+    bo_typ: BoTyp = attr.ib(default=BoTyp.AUSSCHREIBUNG)
+    #: Vom Herausgeber der Ausschreibung vergebene eindeutige Nummer
+    ausschreibungsnummer: str = attr.ib(validator=attr.validators.instance_of(str))
+    #: Aufzählung für die Typisierung von Ausschreibungen
+    ausschreibungstyp: Ausschreibungstyp = attr.ib(validator=attr.validators.instance_of(Ausschreibungstyp))
+    #: Bezeichnungen für die Ausschreibungsphasen
+    ausschreibungsstatus: Ausschreibungsstatus = attr.ib(validator=attr.validators.instance_of(Ausschreibungsstatus))
+
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:37]:
```suggestion
```

File: src/bo4e/bo/ausschreibung.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    #: Kennzeichen, ob die Ausschreibung kostenpflichtig ist
+    kostenpflichtig: bool = attr.ib(validator=attr.validators.instance_of(bool))
+    #: Gibt den Veröffentlichungszeitpunkt der Ausschreibung an
+    veroeffentlichungszeitpunkt: datetime = attr.ib(validator=attr.validators.instance_of(datetime))
+
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:42]: I'm not sure myself what looks nicer here. For the other objects we have always listed the attributes without an empty line.
```suggestion

```

File: src/bo4e/bo/ausschreibung.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    ausschreibender: Geschaeftspartner = attr.ib(validator=attr.validators.instance_of(Geschaeftspartner))
+    """
+    Mit diesem Objekt können Geschäftspartner übertragen werden.
+    Sowohl Unternehmen, als auch Privatpersonen können Geschäftspartner sein
+    """
+
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:48]:
```suggestion
```

File: src/bo4e/bo/ausschreibung.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    #: Die einzelnen Lose, aus denen sich die Ausschreibung zusammensetzt
+    lose: List[Ausschreibungslos] = attr.ib(
+        validator=attr.validators.deep_iterable(
+            member_validator=attr.validators.instance_of(Ausschreibungslos),
+            iterable_validator=check_list_length_at_least_one,
+        )
+    )
+    # optional attributes
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:68]:
```suggestion

    # optional attributes
```

File: src/bo4e/bo/ausschreibung.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    Diese Komponente wird zur Abbildung von Zeiträumen in Form von Dauern oder der Angabe von Start und Ende verwendet.
+    Es muss daher entweder eine Dauer oder ein Zeitraum in Form von Start und Ende angegeben sein
+    """
+
+    bindefrist: Zeitraum = attr.ib(validator=attr.validators.instance_of(Zeitraum))
+    """
+    Diese Komponente wird zur Abbildung von Zeiträumen in Form von Dauern oder der Angabe von Start und Ende verwendet.
+    Es muss daher entweder eine Dauer oder ein Zeitraum in Form von Start und Ende angegeben sein
+    """
+
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:60]:
```suggestion
```

File: src/bo4e/bo/ausschreibung.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    Mit diesem Objekt können Geschäftspartner übertragen werden.
+    Sowohl Unternehmen, als auch Privatpersonen können Geschäftspartner sein
+    """
+
+    abgabefrist: Zeitraum = attr.ib(validator=attr.validators.instance_of(Zeitraum))
+    """
+    Diese Komponente wird zur Abbildung von Zeiträumen in Form von Dauern oder der Angabe von Start und Ende verwendet.
+    Es muss daher entweder eine Dauer oder ein Zeitraum in Form von Start und Ende angegeben sein
+    """
+
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:54]:
```suggestion
```

File: src/bo4e/bo/ausschreibung.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    kostenpflichtig = fields.Bool()
+    veroeffentlichungszeitpunkt = fields.DateTime()
+    ausschreibender = fields.Nested(GeschaeftspartnerSchema)
+    abgabefrist = fields.Nested(ZeitraumSchema)
+    bindefrist = fields.Nested(ZeitraumSchema)
+    lose = fields.List(fields.Nested(AusschreibungslosSchema))
+
+    # optional attributes
+    ausschreibungportal = EnumField(Ausschreibungsportal, load_default=None)
+    webseite = fields.Str(load_default=None)
```
Comment (hf-krechan) [inline src/bo4e/bo/ausschreibung.py:99]: The `load_default` approach will be deprecated in newer marshmallow versions: https://marshmallow.readthedocs.io/en/stable/upgrading.html?highlight=allow_none#upgrading-to-2-0
```suggestion
    ausschreibungportal = EnumField(Ausschreibungsportal, allow_none=True)
    webseite = fields.Str(allow_none=True)
```

File: tests/test_ausschreibung.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                    bindefrist=example_zeitraum,
+                    ausschreibender=Geschaeftspartner(
+                        name1="Batman",
+                        gewerbekennzeichnung=True,
+                        geschaeftspartnerrolle=[Geschaeftspartnerrolle.LIEFERANT],
+                        partneradresse=example_adresse,
+                    ),
+                    lose=[example_ausschreibungslos],
+                )
+            ),
```
Comment (hf-krechan) [inline tests/test_ausschreibung.py:40]: I would recommend to add a minimal example to this test.
Compare to [unittest in BO Rechnung](https://github.com/Hochfrequenz/BO4E-python/blob/0a48a815aef7c61c75856ca97e021ae6a570da10/tests/test_rechnung.py#L83)

**Review by hf-krechan — CHANGES_REQUESTED** [review]

**Review by hf-krechan — APPROVED** [review]

**Pull request merged**